okio/src/test/java/okio/BufferedSourceTest.java

```diff
@@ -810,4 +810,109 @@ private void assertLongDecimalString(String s, long expected) throws IOException
     assertLongDecimalString("-00000000000000009223372036854775808", -9223372036854775808L);
     assertLongDecimalString(TestUtil.repeat('0', Segment.SIZE + 1) + "1", 1);
   }
+
+  @Test public void select() throws IOException {
+    List<ByteString> selection = Arrays.asList(
+        ByteString.encodeUtf8("ROCK"),
+        ByteString.encodeUtf8("SCISSORS"),
+        ByteString.encodeUtf8("PAPER"));
+
+    sink.writeUtf8("PAPER,SCISSORS,ROCK");
+    assertEquals(2, source.select(selection));
+    assertEquals(',', source.readByte());
+    assertEquals(1, source.select(selection));
+    assertEquals(',', source.readByte());
+    assertEquals(0, source.select(selection));
+    assertEquals(true, source.exhausted());
+  }
+
+  @Test public void selectSpanningMultipleSegments() throws IOException {
+    ByteString commonPrefix = TestUtil.randomBytes(Segment.SIZE + 10);
+    ByteString a = new Buffer().write(commonPrefix).writeUtf8("a").readByteString();
+    ByteString bc = new Buffer().write(commonPrefix).writeUtf8("bc").readByteString();
+    ByteString bd = new Buffer().write(commonPrefix).writeUtf8("bd").readByteString();
+    List<ByteString> selection = Arrays.asList(a, bc, bd);
+
+    sink.write(bd);
+    sink.write(a);
+    sink.write(bc);
+
+    assertEquals(2, source.select(selection));
+    assertEquals(0, source.select(selection));
+    assertEquals(1, source.select(selection));
+    assertEquals(true, source.exhausted());
+  }
+
+  @Test public void selectNotFound() throws IOException {
+    List<ByteString> selection = Arrays.asList(
+        ByteString.encodeUtf8("ROCK"),
+        ByteString.encodeUtf8("SCISSORS"),
+        ByteString.encodeUtf8("PAPER"));
+
+    sink.writeUtf8("SPOCK");
+    assertEquals(-1, source.select(selection));
+    assertEquals("SPOCK", source.readUtf8());
+  }
+
+  @Test public void selectValuesHaveCommonPrefix() throws IOException {
+    List<ByteString> selection = Arrays.asList(
+        ByteString.encodeUtf8("abcd"),
+        ByteString.encodeUtf8("abce"),
+        ByteString.encodeUtf8("abcc"));
+
+    sink.writeUtf8("abcc").writeUtf8("abcd").writeUtf8("abce");
+    assertEquals(2, source.select(selection));
+    assertEquals(0, source.select(selection));
+    assertEquals(1, source.select(selection));
+  }
+
+  @Test public void selectLongerThanSource() throws IOException {
+    List<ByteString> selection = Arrays.asList(
+        ByteString.encodeUtf8("abcd"),
+        ByteString.encodeUtf8("abce"),
+        ByteString.encodeUtf8("abcc"));
+    sink.writeUtf8("abc");
+    assertEquals(-1, source.select(selection));
+    assertEquals("abc", source.readUtf8());
+  }
+
+  @Test public void selectReturnsFirstByteStringThatMatches() throws IOException {
+    List<ByteString> selection = Arrays.asList(
+        ByteString.encodeUtf8("abcd"),
+        ByteString.encodeUtf8("abc"),
+        ByteString.encodeUtf8("abcde"));
+    sink.writeUtf8("abcdef");
+    assertEquals(0, source.select(selection));
```

Comment:
Yeah. Doing so would favor a trie or equivalent data structure which knew
that it needed to disambiguate.

On Sat, Apr 16, 2016, 5:02 PM Jesse Wilson <notifications@github.com> wrote:

> In okio/src/test/java/okio/BufferedSourceTest.java
> <https://github.com/square/okio/pull/211#discussion_r59973944>:
>
> > +    List<ByteString> selection = Arrays.asList(
> > +        ByteString.encodeUtf8("abcd"),
> > +        ByteString.encodeUtf8("abce"),
> > +        ByteString.encodeUtf8("abcc"))
>
> ;
> > +    sink.writeUtf8("abc");
> > +    assertEquals(-1, source.select(selection));
> > +    assertEquals("abc", source.readUtf8());
> > +  }
> > +
> > +  @Test public void selectReturnsFirstByteStringThatMatches() throws IOException {
> > +    List<ByteString> selection = Arrays.asList(
> > +        ByteString.encodeUtf8("abcd"),
> > +        ByteString.encodeUtf8("abc"),
> > +        ByteString.encodeUtf8("abcde"));
> > +    sink.writeUtf8("abcdef");
> > +    assertEquals(0, source.select(selection));
>
> Tough to say. Attempting to match a longer string could cause the read to
> block on loading more data. Right now the caller can get whatever behavior
> she wants: sort the choices for shortest match, reverse sort for linear
> match, and arbitrary order if that's preferred.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/square/okio/pull/211/files/c9c14e9923934cefb565803b9a57f69fe206988e#r59973944>
>